tests/scaling_test.go

```diff
@@ -221,37 +227,106 @@ func Test_ScaleToZero(t *testing.T) {
+
+	var loadOutput string
+	var foundReplicas uint64
+	select {
+	case loadOutput = <-output:
+		// end of load
+		cancelLoad()
+		cancelWatch()
+		t.Log("request load finished")
+	case foundReplicas = <-replicas:
+		// reached desired replica count
+		cancelLoad()
+		cancelWatch()
+	}
+
+	if foundReplicas != maxReplicas {
+		t.Logf("function load output %s", loadOutput)
+		t.Fatalf("never reached max scale %d, only %d replicas after %d attempts", maxReplicas, foundReplicas, attempts)
+	}
+
+	// cooldown
+	replicas, cancelWatch = watchReplicaCount(t, functionName, 0)
+	defer cancelWatch()
+
+	select {
+	case <-replicas:
+		return
+	case <-time.After(65 * time.Second):
+		fnc := get(t, functionName)
+		if fnc.AvailableReplicas != 0 {
+			t.Fatalf("got %d replicas, wanted %d", fnc.Replicas, 0)
+		}
+	}
+}
+
+// sendRequestLoad is a helper method that will construct and run the hey Worker in a separate
+// goroutine. When the worker finished, the output will be sent over the channel.  The caller
+// can use the cancel method to stop the worker early
+func sendRequestLoad(t *testing.T, functionName string, loadAttempts int) (_ <-chan string, cancel func()) {
+	results := make(chan string)
+
 	functionURL := gatewayUrl(t, path.Join("function", functionName), "")
 	req, err := http.NewRequest(http.MethodPost, functionURL, nil)
 	if err != nil {
-		t.Fatalf("error with request %s ", err)
+		t.Fatalf("error building function load request %s ", err)
 	}
 
 	var loadOutput bytes.Buffer
-	attempts := 1000
 	functionLoad := requester.Work{
 		Request:           req,
-		N:                 attempts,
+		N:                 loadAttempts,
 		Timeout:           10,
 		C:                 2,
 		QPS:               5.0,
 		DisableKeepAlives: true,
 		Writer:            &loadOutput,
 	}
 
-	functionLoad.Init()
-	functionLoad.Run()
+	go func() {
+		functionLoad.Init()
+		functionLoad.Run()
 
-	fnc := get(t, functionName)
-	if fnc.Replicas != maxReplicas {
-		t.Logf("function load output %s", loadOutput.String())
-		t.Fatalf("never reached max scale %d, only %d replicas after %d attempts", maxReplicas, fnc.Replicas, attempts)
-	}
+		results <- loadOutput.String()
+		close(results)
+	}()
 
-	// cooldown
-	time.Sleep(2 * time.Minute)
-	fnc = get(t, functionName)
-	if fnc.Replicas != 0 {
-		t.Fatalf("got %d replicas, wanted 0", fnc.Replicas)
-	}
+	return results, functionLoad.Stop
+}
+
+// watchReplicaCount is helper method what will use the OpenFaaS API to watch and wait for a
+// function to reach a desired available replica count. Once reached, the count is sent
+// over the result channel. The caller can use the cancel method to stop the replica watcher early.
+func watchReplicaCount(t *testing.T, functionName string, target uint64) (_ <-chan uint64, cancel func()) {
+	t.Helper()
+	results := make(chan uint64)
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	go func(t *testing.T) {
+		defer close(results)
+		start := time.Now()
+
+		for {
```

Comment:
Would TimerFunc or similar make more sense here or an interval? It seems rather verbose.

Reply:
What is TimerFunc, i don't see it in the stdlib

Reply:
The current design isn't verbose because of the time, even if i remove all of the timing components, the verbosity is so that i can pass back a `cancel` method and also watch for that cancel to be called, in this section of the code https://github.com/openfaas/certifier/pull/34/files#diff-7e0b111256bb46309f2975b5947e4f50R237-R246

This allows the caller to control the timeout or in our case, cancel if the load test is done

Reply:
https://golang.org/pkg/time/#Timer

AfterFunc or NewTimer might be less verbose.

Reply:
There's other related functions available too depending on requirements https://golang.org/pkg/time/#After

Reply:
neither of those are used to time calls. `After` calls the function after the duration has passed. Similarly, `Timer` creates a channel that will receive the current time _after_ a specified duration has passed.  

Both of these methods you mention are about triggering something after a set amount of time. This `start:=time.Now()` is about recording how much time has elapsed. 

If the issue is with the context with cancel. Neither of the methods you provide actually allow the caller to control the cancel.  The current design allows me to provide the `cancel` method to the original caller so that it can decide when to call the cleanup instead of forcing the caller to wait for a specific time period to elapse.